### PR TITLE
CSS Transforms Module Improvements

### PR DESCRIPTION
**Pull request by zfox23**
### Description

1. Improves the existing CSS Transforms module with a new interactive example and additional prose for context.

I'm not sure what the best way is to see and test the new `css-examples` page. One way is to download its source locally, then open it in your browser: [`transforms.html`](https://raw.githubusercontent.com/mdn/css-examples/c462161f2838669eff5e64cc2adc41c51dd6bf44/modules/transforms.html )

### Motivation

Main motivation: Improving our documentation as a part of [the Interop 2023 project](https://github.com/mdn/mdn/issues/416).

### Related issues and pull requests

- `content` PR (for additional module prose): https://github.com/mdn/content/pull/30783
- `css-examples` PR (for the module page's "CSS transforms in action" interactive example section): https://github.com/mdn/css-examples/pull/160

Once these two PRs are merged, https://github.com/mdn/mdn/issues/488 will be resolved.

### CODE REVIEW

File: files/en-us/web/css/css_transforms/styling_animated_buttons_with_css_transforms/index.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
+---
+title: Styling animated buttons with CSS transforms
+slug: Web/CSS/CSS_transforms/styling_animated_buttons_with_css_transforms
```
Comment (teoli2003) [inline files/en-us/web/css/css_transforms/styling_animated_buttons_with_css_transforms/index.md:3]: Drive-by comment, slugs should be capitalized consistently:
```suggestion
slug: Web/CSS/CSS_transforms/Styling_animated_buttons_with_css_transforms
```

Comment (zfox23) [reply files/en-us/web/css/css_transforms/styling_animated_buttons_with_css_transforms/index.md]: thank you 😊

File: files/en-us/web/css/css_transforms/index.md [hunk 2]
```diff
@@ -18,12 +32,18 @@ The **CSS transforms** module defines how elements styled with CSS can be transf
...
+- [`skew`](/en-US/docs/Web/CSS/transform-function/skew)
 - {{cssxref("transform")}}
 - {{cssxref("transform-box")}}
 - {{cssxref("transform-origin")}}
 - {{cssxref("transform-style")}}
 - {{cssxref("translate")}}
 
+### Functions
+
+- See {{cssxref("&lt;transform-function&gt;")}}
```
Comment (estelle) [inline files/en-us/web/css/css_transforms/index.md:44]: Every function needs to all be listed here. CSS module landing page links to all the features of the specification. Everything in the spec should presented as a bulleted list under the relevant headings.


```suggestion
- {{cssxref("matrix()")}}
- {{cssxref("rotate()")}}
- {{cssxref("scale()")}}
- {{cssxref("scaleX()")}}
- {{cssxref("scaleY()")}}
- {{cssxref("skew()")}}
- {{cssxref("skewX()")}}
- {{cssxref("skewY()")}}
- {{cssxref("translate()")}}
- {{cssxref("translateX()")}}
- {{cssxref("translateY()")}}
```
then add the functions listed here:

matrix3d() = matrix3d( [<number>](https://drafts.csswg.org/css-values-4/#number-value)[#{16}](https://drafts.csswg.org/css-values-4/#mult-comma) )
specifies a 3D transformation as a 4x4 homogeneous matrix of 16 values in column-major order.

translate3d() = translate3d( [<length-percentage>](https://drafts.csswg.org/css-values-4/#typedef-length-percentage) [,](https://drafts.csswg.org/css-values-4/#comb-comma) <length-percentage> , [<length>](https://drafts.csswg.org/css-values-4/#length-value) )
specifies a [3D translation](https://drafts.csswg.org/css-transforms-2/#Translate3dDefined) by the vector [tx,ty,tz], with tx, ty and tz being the first, second and third translation-value parameters respectively.

translateZ() = translateZ( [<length>](https://drafts.csswg.org/css-values-4/#length-value) )
specifies a [3D translation](https://drafts.csswg.org/css-transforms-2/#Translate3dDefined) by the vector [0,0,tz] with the given amount in the Z direction.

scale3d() = scale3d( [ [<number>](https://drafts.csswg.org/css-values-4/#number-value) [|](https://drafts.csswg.org/css-values-4/#comb-one) [<percentage>](https://drafts.csswg.org/css-values-4/#percentage-value) ][#{3}](https://drafts.csswg.org/css-values-4/#mult-comma) )
specifies a [3D scale](https://drafts.csswg.org/css-transforms-2/#Scale3dDefined) operation by the [sx,sy,sz] scaling vector described by the 3 parameters.

scaleZ() = scaleZ( [ [<number>](https://drafts.csswg.org/css-values-4/#number-value) [|](https://drafts.csswg.org/css-values-4/#comb-one) [<percentage>](https://drafts.csswg.org/css-values-4/#percentage-value) ] )
specifies a [3D scale](https://drafts.csswg.org/css-transforms-2/#Scale3dDefined) operation using the [1,1,sz] scaling vector, where sz is given as the parameter.

rotate3d() = rotate3d( [<number>](https://drafts.csswg.org/css-values-4/#number-value) [,](https://drafts.csswg.org/css-values-4/#comb-comma) <number> , <number> , [ [<angle>](https://drafts.csswg.org/css-values-4/#angle-value) [|](https://drafts.csswg.org/css-values-4/#comb-one) [<zero>](https://drafts.csswg.org/css-values-4/#zero-value) ] )
specifies a [3D rotation](https://drafts.csswg.org/css-transforms-2/#Rotate3dDefined) by the angle specified in last parameter about the [x,y,z] direction vector described by the first three parameters. A direction vector that cannot be normalized, such as [0,0,0], will cause the rotation to not be applied.

NOTE: the rotation is clockwise as one looks from the end of the vector toward the origin.

rotateX() = rotateX( [ [<angle>](https://drafts.csswg.org/css-values-4/#angle-value) [|](https://drafts.csswg.org/css-values-4/#comb-one) [<zero>](https://drafts.csswg.org/css-values-4/#zero-value) ] )
same as rotate3d(1, 0, 0, <angle>).

rotateY() = rotateY( [ [<angle>](https://drafts.csswg.org/css-values-4/#angle-value) [|](https://drafts.csswg.org/css-values-4/#comb-one) [<zero>](https://drafts.csswg.org/css-values-4/#zero-value) ] )
same as rotate3d(0, 1, 0, <angle>).

rotateZ() = rotateZ( [ [<angle>](https://drafts.csswg.org/css-values-4/#angle-value) [|](https://drafts.csswg.org/css-values-4/#comb-one) [<zero>](https://drafts.csswg.org/css-values-4/#zero-value) ] )
same as rotate3d(0, 0, 1, <angle>), which is a 3d transform equivalent to the 2d transform rotate(<angle>).

perspective()

we also need the SVG `transform` attribute.

File: files/en-us/web/css/css_transforms/index.md [hunk 3]
```diff
@@ -18,12 +32,18 @@ The **CSS transforms** module defines how elements styled with CSS can be transf
 - {{cssxref("perspective-origin")}}
 - {{cssxref("rotate")}}
 - {{cssxref("scale")}}
+- [`skew`](/en-US/docs/Web/CSS/transform-function/skew)
```
Comment (estelle) [inline files/en-us/web/css/css_transforms/index.md:35]: this is not a property. it's a function.

The only properties defined in this module are transform, transform-box, and transform-origin. The spec needs to be updated to https://drafts.csswg.org/css-transforms-2 instead. 

this list should be:

perspective
perspective-origin
rotate
scale
transform
t-box
t-origin
t-style
translate
transform
transform-origin
transformable element
transformation matrix
transformed element

File: files/en-us/web/css/css_transforms/index.md [hunk 4]
```diff
@@ -32,6 +52,17 @@ The **CSS transforms** module defines how elements styled with CSS can be transf
 
 - [Using CSS transforms](/en-US/docs/Web/CSS/CSS_transforms/Using_CSS_transforms)
   - : Step-by-step tutorial about how to transform elements styled with CSS.
+- [Styling animated buttons with CSS transforms](/en-US/docs/Web/CSS/CSS_transforms/styling_animated_buttons_with_css_transforms)
+  - : Contains several examples of `<button>` elements styled using animated CSS transforms.
+
+## Related concepts
+
```
Comment (estelle) [inline files/en-us/web/css/css_transforms/index.md:59]: Please take a look at the module landing pages that have been updated and conform to how those are presented. 

These are useful links, but should be presented within guides, related concepts as list items under their module as a header, or in See also.

There are a ton of related concepts not listed. A good place to figure out what to include is the

**Review by estelle — CHANGES_REQUESTED** [review]
I only reviewed part of the css_transforms pages. The page needs to follow the css-module landing page template.

**Comment by zfox23** [comment]
Thanks for your feedback @estelle. It should all be addressed now. A few things:

1. I think the "using CSS transforms to enable GPU acceleration" guide should stay, based on [community feedback](https://discord.com/channels/1009925603572600863/1179814011081076799/1179825210141196318) and due to this quick guide including example code.
2. I have deleted the 3-example-button page.
3. A few of the properties you mentioned, including `transform-box` and `transform-origin`, are not present in the CSS2 specification. I have included them in [#properties](https://pr30783.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/CSS_transforms#properties) with the label "_(Defined in [Level 1](https://drafts.csswg.org/css-transforms-1/))_". Please advise if you'd prefer them somewhere else.
4. I did my best with the translation/migration of terms defined in the CSS2 spec to terms defined/linked to on this page. Please advise if you believe I missed any.

**Review by teoli2003 — DISMISSED** [review]
I was about to post this yesterday, but Estelle was quicker than me, so I dismissed it. But here is my position:

I have **strong** concerns about a standalone _Forcing GPU acceleration with CSS transforms_.

This trick has to be documented, but having it on its own page is **harmful**. Why? Because this is a dirty trick and not a good practice. What's the difference? A _dirty trick_ is something needed as a last resort, circumventing a limitation of the platform, while a _good practice_ is an esthetically pleasing way of implementing a pattern.

We want to document dirty tricks that are still relevant, so we want to speak about how to force GPU acceleration. Nevertheless, we want it to be clear that this is something to use as a last resort and not to add it to every animation or transition.

That is why we must speak about it only _in context_ as the last action to speed up an animation. What we need is a page explaining how to speed up animations (and transitions) that would have at least the following sections:

- How to detect an animation speed problem (like how do we know if frames are dropped).
- How to prevent _layout_ to be recalculated/recomposited (avoiding moving using absolute `top`, `right` and using `translate`)
- How to prevent unnecessary _repaint_ by using transformation like `blur()` and shadows cautiously.
- … ( I expect information about the `will-change` property too)
- As a last resort, a section about trying to force GPU, explaining the caveats (more memory use, may not work everywhere, may disappear in the future as it is tied to implementation).

Also, I think it may be an excellent opportunity to fuse (or consider the fusion) these two pages with this new one:

-  https://developer.mozilla.org/en-US/docs/Web/Performance/Animation_performance_and_frame_rate
-  https://developer.mozilla.org/en-US/docs/Web/Performance/CSS_JavaScript_animation_performance

I have no strong feelings about where the page should live, but it should not be under CSS Transitions as it is not specific to it. I would put it under Web/Performance until we have a Diátaxis-compliant information system for guides as considered.

As a standalone page, without the context and the other information, it is a big no for me as it would harm web developers.

**Comment by chrisdavidmills** [comment]
+1 on what @teoli2003 and @estelle said (they beat me to it ;-) ). The information as-is is not terrible, pending the needed clarification and context settings mentioned in the othe comments. However, it seems like a very simplified view. I'm pretty sure this is not the only way to force HW acceleration, it would benefit from a discussion of when you should do this, and alternative approaches the reader could consider.

It might also be nice to explain why the HW acceleration trick works; talk about the separate composite layers, etc.

Also, the example is pretty trivial; maybe build slightly something slightly more expansive that shows a lot of animation going on, maybe several DOM elements in two different boxes, with a complex background gradient, one with HW acceleration and one without, so the reader can examine the difference. Also provide a button to toggle animation start/stop.

**Comment by zfox23** [comment]
hi folks. most recently, I have removed both of the new guides I wrote from this pr. i would like to get what's present in this pr merged when possible. could someone please review what's left? thank you.

File: files/en-us/web/css/css_transforms/index.md [hunk 5]
```diff
@@ -2,13 +2,31 @@
 title: CSS transforms
 slug: Web/CSS/CSS_transforms
 page-type: css-module
-spec-urls: https://drafts.csswg.org/css-transforms/
+spec-urls: https://drafts.csswg.org/css-transforms-2/
```
Comment (estelle) [inline files/en-us/web/css/css_transforms/index.md:5]:
```suggestion
spec-urls: 
  - https://drafts.csswg.org/css-transforms/
  - https://drafts.csswg.org/css-transforms-2/
```

File: files/en-us/web/css/css_transforms/index.md [hunk 6]
```diff
@@ -19,10 +37,35 @@ The **CSS transforms** module defines how elements styled with CSS can be transf
 - {{cssxref("rotate")}}
 - {{cssxref("scale")}}
 - {{cssxref("transform")}}
-- {{cssxref("transform-box")}}
-- {{cssxref("transform-origin")}}
+- {{cssxref("transform-box")}} _(Defined in [Level 1](https://drafts.csswg.org/css-transforms-1/))_
+- {{cssxref("transform-origin")}} _(Defined in [Level 1](https://drafts.csswg.org/css-transforms-1/))_
```
Comment (estelle) [inline files/en-us/web/css/css_transforms/index.md:41]:
```suggestion
- {{cssxref("transform-box")}}
- {{cssxref("transform-origin")}}
```

File: files/en-us/web/css/css_transforms/index.md [hunk 7]
```diff
@@ -32,7 +75,46 @@ The **CSS transforms** module defines how elements styled with CSS can be transf
...
   - : Step-by-step tutorial about how to transform elements styled with CSS.
+- [Coordinate systems](/en-US/docs/Web/CSS/CSSOM_view/Coordinate_systems)
+  - : Describes the way pixel locations are defined in the CSS object model.
+- [Matrix math for the web](/en-US/docs/Web/API/WebGL_API/Matrix_math_for_the_web)
+  - : Describes how object transformations can be represented by mathematical matrices.
+
+## Related concepts
+
+- CSS Properties:
+  - [`background-position`](/en-US/docs/Web/CSS/background-position)
```
Comment (estelle) [inline files/en-us/web/css/css_transforms/index.md:86]:
```suggestion
  - [`animation`](/en-US/docs/Web/CSS/animation)
  - [`background-position`](/en-US/docs/Web/CSS/background-position)
```

File: files/en-us/web/css/css_transforms/index.md [hunk 8]
```diff
@@ -32,7 +75,46 @@ The **CSS transforms** module defines how elements styled with CSS can be transf
...
+  - [`contain`](/en-US/docs/Web/CSS/contain)
+  - [`content-visibility`](/en-US/docs/Web/CSS/content-visibility)
+  - [`isolation`](/en-US/docs/Web/CSS/isolation)
+  - [`mask`](/en-US/docs/Web/CSS/mask)
+  - [`mask-border-source`](/en-US/docs/Web/CSS/mask-border-source)
+  - [`mask-image`](/en-US/docs/Web/CSS/mask-image)
+  - [`mix-blend-mode`](/en-US/docs/Web/CSS/mix-blend-mode)
+  - [`opacity`](/en-US/docs/Web/CSS/opacity)
+  - [`overflow`](/en-US/docs/Web/CSS/overflow)
+  - [`visibility`](/en-US/docs/Web/CSS/visibility)
```
Comment (estelle) [inline files/en-us/web/css/css_transforms/index.md:98]:
```suggestion
  - [`transition`](/en-US/docs/Web/CSS/transition)
  - [`visibility`](/en-US/docs/Web/CSS/visibility)
```

File: files/en-us/web/css/css_transforms/index.md [hunk 9]
```diff
@@ -32,7 +75,46 @@ The **CSS transforms** module defines how elements styled with CSS can be transf
...
+  - [`<set>`](/en-US/docs/Web/SVG/Element/set)
+  - [`transform`](/en-US/docs/Web/SVG/Attribute/transform)
 
 ## Specifications
 
 {{Specifications}}
+
+## See also
+
+- CSS properties: [`animation`](/en-US/docs/Web/CSS/animation), [`transition`](/en-US/docs/Web/CSS/transition)
```
Comment (estelle) [inline files/en-us/web/css/css_transforms/index.md:120]:
```suggestion
- [CSS animations](/en-US/docs/Web/CSS/CSS_animations) module
- [CSS transitions](/en-US/docs/Web/CSS/CSS_transitions) module
```

File: files/en-us/web/css/css_transforms/index.md [hunk 10]
```diff
@@ -32,7 +75,46 @@ The **CSS transforms** module defines how elements styled with CSS can be transf
...
+  - [`<number>`](/en-US/docs/Web/CSS/number)
+  - [`<percentage>`](/en-US/docs/Web/CSS/percentage)
+  - [`<position>`](/en-US/docs/Web/CSS/position_value)
+- Glossary terms:
+  - [Interpolation](/en-US/docs/Glossary/Interpolation)
+  - [Stacking context](/en-US/docs/Glossary/Stacking_context)
+- SVG concepts:
+  - [`<animate>`](/en-US/docs/Web/SVG/Element/animate)
+  - [`<set>`](/en-US/docs/Web/SVG/Element/set)
+  - [`transform`](/en-US/docs/Web/SVG/Attribute/transform)
```
Comment (estelle) [inline files/en-us/web/css/css_transforms/index.md:112]:
```suggestion
  - [`<animate>`](/en-US/docs/Web/SVG/Element/animate) element
  - [`<set>`](/en-US/docs/Web/SVG/Element/set) element
  - [`transform`](/en-US/docs/Web/SVG/Attribute/transform) attribute
```

Comment (estelle) [reply files/en-us/web/css/css_transforms/index.md:112]: did you want to add https://developer.mozilla.org/en-US/docs/Web/SVG/Element/animateTransform?

File: files/en-us/web/css/css_transforms/index.md [hunk 11]
```diff
@@ -32,7 +75,46 @@ The **CSS transforms** module defines how elements styled with CSS can be transf
 
 - [Using CSS transforms](/en-US/docs/Web/CSS/CSS_transforms/Using_CSS_transforms)
   - : Step-by-step tutorial about how to transform elements styled with CSS.
+- [Coordinate systems](/en-US/docs/Web/CSS/CSSOM_view/Coordinate_systems)
+  - : Describes the way pixel locations are defined in the CSS object model.
+- [Matrix math for the web](/en-US/docs/Web/API/WebGL_API/Matrix_math_for_the_web)
```
Comment (estelle) [inline files/en-us/web/css/css_transforms/index.md:80]:
```suggestion
- [Perfomance fundamentals: Using CSS transforms](/en-US/docs/Web/Performance/Fundamentals#use_css_transforms)
   - : An overview of web performance fundamentals, including how CSS tranforms can help.
- [Matrix math for the web](/en-US/docs/Web/API/WebGL_API/Matrix_math_for_the_web)
```

File: files/en-us/web/css/css_transforms/index.md (hunk 9 above)
Comment (estelle) [reply files/en-us/web/css/css_transforms/index.md:120]: let's also link to SVG transforms overview, if there is one, and maybe link the SVG in "SVG concepts" to the svg landing page.

**Review by estelle — COMMENTED** [review]
Getting there. i made a few suggestions for content tweaks. You may want to reword the guide one ;). 

I think two spec urls are possible. I copied the syntax from the cascade and colors modules.

File: files/en-us/web/css/css_transforms/index.md (hunk 10 above)
Comment (zfox23) [reply files/en-us/web/css/css_transforms/index.md:116]: sure!

File: files/en-us/web/css/css_transforms/index.md [hunk 12]
```diff
@@ -2,13 +2,33 @@
...
+
+The [`perspective-origin`](/en-US/docs/Web/CSS/perspective-origin) sliders determine where you, the viewer, are looking into the 3D space for purposes of determining the view's _vanishing point_. This vanishing point is indicated by a small red dot. You can imagine modifying these sliders as physically moving your head up, down, left, and right to see different parts of the cube without moving the cube itself.
+
+The `backface-visibility` checkbox determines whether the cube's back faces are set to `visible` or `hidden`.
+
+The cube in the above example is comprised of six `<div>` elements, all of which are styled with CSS to create the cube's faces. The cube is not drawn using a 2D or 3D canvas context, so **you can inspect the faces of the cube with your browser's developer tools as you would inspect any other DOM element**. Try using your browser's developer tools element picker to inspect different faces of the cube as you transform its position and rotation.
+
+> **Note:** that the order in which 3D rotations are applied affects the resultant transformation. In the above example, rotations are applied in the order X → Y → Z.
```
Comment (estelle) [inline files/en-us/web/css/css_transforms/index.md]:
```suggestion
> **Note:** The order in which transformations, including 3D rotations, are applied affects the resultant transformation. In the above example, transforms are translated, scaled, rotated, then skewed. The rotations are applied in the order X → Y → Z.
```

**Review by estelle — APPROVED** [review]
just one last tweak. Approved!

File: files/en-us/web/css/css_transforms/index.md [hunk 13]
```diff
@@ -2,13 +2,33 @@
...
+
+The [`perspective-origin`](/en-US/docs/Web/CSS/perspective-origin) sliders determine where you, the viewer, are looking into the 3D space for purposes of determining the view's _vanishing point_. This vanishing point is indicated by a small red dot. You can imagine modifying these sliders as physically moving your head up, down, left, and right to see different parts of the cube without moving the cube itself.
+
+The `backface-visibility` checkbox determines whether the cube's back faces are set to `visible` or `hidden`.
+
+The cube in the above example is comprised of six `<div>` elements, all of which are styled with CSS to create the cube's faces. The cube is not drawn using a 2D or 3D canvas context, so **you can inspect the faces of the cube with your browser's developer tools as you would inspect any other DOM element**. Try using your browser's developer tools element picker to inspect different faces of the cube as you transform its position and rotation.
+
+> **Note:** The order in which 3D rotations are applied affects the resultant transformation. In the above example, rotations are applied in the order X → Y → Z.
```
Comment (estelle) [inline files/en-us/web/css/css_transforms/index.md:28]:
```suggestion
> **Note:** The order in which transformations, including 3D rotations, are applied affects the resultant transformation. In the above example, transforms are translated, scaled, rotated, then skewed. The rotations are applied in the order X → Y → Z.
```

File: files/en-us/web/css/css_transforms/index.md (hunk 12 above)
Comment (zfox23) [reply files/en-us/web/css/css_transforms/index.md]: thank you for your suggestion. in the future, i would appreciate an opportunity to review suggestions before you commit them and/or merge the PR (except in cases where we'd talked about that being OK previously).

Comment (estelle) [reply files/en-us/web/css/css_transforms/index.md]: Will do.
